Leanbootstrap

dmbates opened the pull request:
- Addresses #247 

- Avoid keeping a copy of the fitted model in the `MixedModelBootstrap` type.  Just keep enough information to be able to construct the σ and ρ values from the θ values.

- Change the form of the extracted values from a rowtable of recursive `NamedTuples` to a "tidy" dataframe.

dmbates commented:
I would appreciate comments on this PR from the usual suspects (@palday, @debruine, @Nosferican).  Lisa reminded me of tidy data principles whereby all the responses are in a single column with the various classifications as categorical covariates. 

In previous incarnations of this the extracted values of the coefficients, for example, were returned as a vector of NamedTuples on which one had to do a certain amount of confusing to extract the actual numbers.  In the initial version of this PR I add a function called `MixedModels.tidyβ` which returns a RowTable of iteration, coefname, and value.  These can easily be converted to a `DataFrame` and summarized or plotted using `by`.  It looks like
```julia
julia> using MixedModels, Random, DataFrames
[ Info: Precompiling MixedModels [ff71e718-51f3-5ec2-a782-8ffcbfa3c316]

julia> sleepstudy = MixedModels.dataset(:sleepstudy);

julia> m1 = fit(MixedModel, @formula(reaction ~ 1+days+(1+days|subj)), sleepstudy);

julia> bsamp = parametricbootstrap(MersenneTwister(6345789), 1000, m1,  use_threads=true);
Progress: 100%|█████████████████████████████████████████████████████████████████████████████████████████████| Time: 0:00:00

julia> β = MixedModels.tidyβ(bsamp)
2000-element Array{NamedTuple{(:iter, :coefname, :β),Tuple{Int64,Symbol,Float64}},1}:
 (iter = 1, coefname = Symbol("(Intercept)"), β = 259.65270368521857)  
 (iter = 1, coefname = :days, β = 12.287978077187747)                  
 (iter = 2, coefname = Symbol("(Intercept)"), β = 256.63351407080955)  
 (iter = 2, coefname = :days, β = 9.271939682932558)                   
 (iter = 3, coefname = Symbol("(Intercept)"), β = 246.55175497118435)  
 (iter = 3, coefname = :days, β = 9.823154966419837)                   
 (iter = 4, coefname = Symbol("(Intercept)"), β = 247.71164723942076)  
 (iter = 4, coefname = :days, β = 10.758121245219636)                  
 (iter = 5, coefname = Symbol("(Intercept)"), β = 252.9927237092217)   
 ⋮                                                                     
 (iter = 997, coefname = Symbol("(Intercept)"), β = 252.27825552296213)
 (iter = 997, coefname = :days, β = 8.86537672584197)                  
 (iter = 998, coefname = Symbol("(Intercept)"), β = 247.74993172101642)
 (iter = 998, coefname = :days, β = 10.93247550866081)                 
 (iter = 999, coefname = Symbol("(Intercept)"), β = 251.4514815775677) 
 (iter = 999, coefname = :days, β = 11.161821621287807)                
 (iter = 1000, coefname = Symbol("(Intercept)"), β = 252.5758822723428)
 (iter = 1000, coefname = :days, β = 8.636175408523435)                

julia> DataFrame(β)
2000×3 DataFrame
│ Row  │ iter  │ coefname    │ β       │
│      │ Int64 │ Symbol      │ Float64 │
├──────┼───────┼─────────────┼─────────┤
│ 1    │ 1     │ (Intercept) │ 259.653 │
│ 2    │ 1     │ days        │ 12.288  │
│ 3    │ 2     │ (Intercept) │ 256.634 │
│ 4    │ 2     │ days        │ 9.27194 │
│ 5    │ 3     │ (Intercept) │ 246.552 │
│ 6    │ 3     │ days        │ 9.82315 │
⋮
│ 1994 │ 997   │ days        │ 8.86538 │
│ 1995 │ 998   │ (Intercept) │ 247.75  │
│ 1996 │ 998   │ days        │ 10.9325 │
│ 1997 │ 999   │ (Intercept) │ 251.451 │
│ 1998 │ 999   │ days        │ 11.1618 │
│ 1999 │ 1000  │ (Intercept) │ 252.576 │
│ 2000 │ 1000  │ days        │ 8.63618 │

julia> by(DataFrame(β), :coefname, extrema = :β => extrema)
2×2 DataFrame
│ Row │ coefname    │ extrema            │
│     │ Symbol      │ Tuple…             │
├─────┼─────────────┼────────────────────┤
│ 1   │ (Intercept) │ (233.471, 274.782) │
│ 2   │ days        │ (4.6929, 15.9774)  │

julia> by(DataFrame(β), :coefname, interval = :β => shortestcovint)
2×2 DataFrame
│ Row │ coefname    │ interval           │
│     │ Symbol      │ Tuple…             │
├─────┼─────────────┼────────────────────┤
│ 1   │ (Intercept) │ (239.407, 264.869) │
│ 2   │ days        │ (7.48773, 13.5335) │
```

Do you think this should be the default extractor?

I like the way the information is stored in a `MixedModelBootstrap` because it is very concise but it is a bit of pain to explain how to extract the relevant values.

palday commented:
I think this is a good compromise solution. We keep the performance-optimized machinery we have, but still provide an object that's convenient for manipulation. Should `tidyβ` also be a property of `MixedModelsBootstrap`? 

I'm guessing `tidy_sigma` and `tidy_sigma_rho` are planned? Those would require a bit more work to get shaped up, but having tidy variants of those would make some visualizations about how singular models behave much more apparent.

dmbates commented:
I was planning on replacing the action of the β property with a call to `tidyβ`.  Make the result that is easier to work with the one that is returned by the simple property extractor.

I haven't yet written the extractors for the sigma and sigma/rho properties.  I wanted to get some feedback on the approach before attacking those.

palday commented:
I like that approach. Anybody who could deal with the old route to get a single beta, can certainly deal with getting a single beta from a row table.

debruine commented:
It's worth looking at how [broom.mixed](https://github.com/bbolker/broom.mixed) organises a tidy table in R

```
require(lme4)
require(broom.mixed)

dat <- lme4::sleepstudy
m <- lmer(Reaction ~ 1 + Days +  (1 + Days | Subject), data = dat)
tidy(m)
```

produces:

```
  effect   group   term            estimate std.error statistic
  <chr>    <chr>   <chr>              <dbl>     <dbl>     <dbl>
1 fixed    NA      (Intercept)     251.          6.82     36.8 
2 fixed    NA      Days             10.5         1.55      6.77
3 ran_pars Subject sd__(Intercept)  24.7        NA        NA   
4 ran_pars Subject sd__Days          5.92       NA        NA   
5 ran_pars Subject cor__(Intercep…   0.0656     NA        NA   
6 ran_pars Residu… sd__Observation  25.6        NA        NA 
```

(or you can set `effects` to `"fixed"` or `"ran_pars"` to just get those)

I'm not advocating doing it that way necessarily (I hate how the term is specified for ran_pars) but if you want to conver lme4 R users, it might be good to give them a familiar output.

kliegl commented:
If I may insert myself: I would prefer the long format Doug proposed. The tidy(m) output requires another transformation from wide to long.

dmbates commented:
I have pushed a couple of commits that allow for extracting the coefficient estimates and the estimates of the standard deviations of the random effects.  I will illustrate using `shortestcovint`.  A similar approach could be used to generate, say, density plots.
```julia
julia> using DataFrames, MixedModels, Random
[ Info: Precompiling DataFrames [a93c6f00-e57d-5684-b7b6-d8193f3e46c0]
[ Info: Precompiling MixedModels [ff71e718-51f3-5ec2-a782-8ffcbfa3c316]

julia> kb07 = MixedModels.dataset(:kb07);

julia> contr = Dict(n => HelmertCoding() for n in (:prec, :load, :spkr));

julia> f2 = @formula rt_trunc ~ 1+spkr*prec*load + (1+prec|item) + (1|subj);

julia> m2 = fit(MixedModel, f2, kb07, contrasts=contr)
Linear mixed model fit by maximum likelihood
 rt_trunc ~ 1 + spkr + prec + load + spkr & prec + spkr & load + prec & load + spkr & prec & load + (1 + prec | item) + (1 | subj)
     logLik        -2 logLik          AIC             BIC       
 -1.43292433×10⁴  2.86584867×10⁴  2.86844867×10⁴   2.8755849×10⁴

Variance components:
             Column      Variance   Std.Dev.   Corr.
item     (Intercept)     133026.917 364.72855
         prec: maintain   63841.496 252.66875 -0.70
subj     (Intercept)      88870.080 298.11085
Residual                 460948.432 678.93183
 Number of obs: 1789; levels of grouping factors: 32, 56

  Fixed-effects parameters:
───────────────────────────────────────────────────────────────────────────────
                                          Estimate  Std.Error  z value  P(>|z|)
───────────────────────────────────────────────────────────────────────────────
(Intercept)                             2181.76       77.4709    28.16   <1e-99
spkr: old                                 67.8114     16.0526     4.22   <1e-4 
prec: maintain                          -333.858      47.4629    -7.03   <1e-11
load: yes                                 78.6849     16.0525     4.90   <1e-6 
spkr: old & prec: maintain               -21.8802     16.0525    -1.36   0.1729
spkr: old & load: yes                     18.3214     16.0526     1.14   0.2537
prec: maintain & load: yes                 4.47097    16.0526     0.28   0.7806
spkr: old & prec: maintain & load: yes    23.5219     16.0525     1.47   0.1428
───────────────────────────────────────────────────────────────────────────────

julia> kbsamp = parametricbootstrap(MersenneTwister(1234321), 10_000, m2, use_threads=true);
Progress: 100%|█████████████████████████████████████████████████████████████████████████████████████████████| Time: 0:00:56

julia> by(DataFrame(kbsamp.β), :coefname, interval = :β => shortestcovint)
8×2 DataFrame
│ Row │ coefname                               │ interval            │
│     │ Symbol                                 │ Tuple…              │
├─────┼────────────────────────────────────────┼─────────────────────┤
│ 1   │ (Intercept)                            │ (2026.75, 2330.41)  │
│ 2   │ spkr: old                              │ (36.994, 98.2024)   │
│ 3   │ prec: maintain                         │ (-428.47, -242.548) │
│ 4   │ load: yes                              │ (47.9728, 111.585)  │
│ 5   │ spkr: old & prec: maintain             │ (-53.4025, 9.36747) │
│ 6   │ spkr: old & load: yes                  │ (-12.0112, 49.9202) │
│ 7   │ prec: maintain & load: yes             │ (-25.7247, 36.8545) │
│ 8   │ spkr: old & prec: maintain & load: yes │ (-6.19643, 55.7169) │

julia> by(DataFrame(kbsamp.σs), [:group, :column], interval = :σ => shortestcovint)
3×3 DataFrame
│ Row │ group  │ column         │ interval           │
│     │ Symbol │ Symbol         │ Tuple…             │
├─────┼────────┼────────────────┼────────────────────┤
│ 1   │ item   │ (Intercept)    │ (265.52, 456.997)  │
│ 2   │ item   │ prec: maintain │ (177.033, 318.248) │
│ 3   │ subj   │ (Intercept)    │ (230.319, 357.947) │

julia> shortestcovint(kbsamp.σ)
(655.0049649114235, 700.5270246634653)
```

I will also generate a σρs property that will have columns of `iter`, `type`, `group`, `column` and `row` where `type` will be `sd` or `corr`, `group` will be the grouping factor, `column` and `row` will be names from the vector-valued random effects for `corr` but `row` will be missing for an `sd` entry.

Does this approach seem reasonable?

dmbates commented:
The reason that I return a `RowTable` that must be explicitly converted to a `DataFrame` is because I don't want `MixedModels` to depend on `DataFrames`, which is not a lightweight dependency.

palday commented:
I think it sounds quite nice, but maybe @debruine has opinions about / suggestions for the naming of `column` and `row`.

dmbates commented:
Yet another version - this time with an `allpars` extractor (a bit of a misnomer because I don't have the correlations in there but I will add them).
```julia
julia> using DataFrames, MixedModels, Random
[ Info: Precompiling MixedModels [ff71e718-51f3-5ec2-a782-8ffcbfa3c316]

julia> kb07 = MixedModels.dataset(:kb07);

julia> contr = Dict(n => HelmertCoding() for n in (:prec, :load, :spkr));

julia> f2 = @formula rt_trunc ~ 1+spkr*prec*load + (1+prec|item) + (1|subj);

julia> m2 = fit(MixedModel, f2, kb07, contrasts=contr)
Linear mixed model fit by maximum likelihood
 rt_trunc ~ 1 + spkr + prec + load + spkr & prec + spkr & load + prec & load + spkr & prec & load + (1 + prec | item) + (1 | subj)
     logLik        -2 logLik          AIC             BIC       
 -1.43292433×10⁴  2.86584867×10⁴  2.86844867×10⁴   2.8755849×10⁴

Variance components:
             Column      Variance   Std.Dev.   Corr.
item     (Intercept)     133026.917 364.72855
         prec: maintain   63841.496 252.66875 -0.70
subj     (Intercept)      88870.080 298.11085
Residual                 460948.432 678.93183
 Number of obs: 1789; levels of grouping factors: 32, 56

  Fixed-effects parameters:
───────────────────────────────────────────────────────────────────────────────
                                          Estimate  Std.Error  z value  P(>|z|)
───────────────────────────────────────────────────────────────────────────────
(Intercept)                             2181.76       77.4709    28.16   <1e-99
spkr: old                                 67.8114     16.0526     4.22   <1e-4 
prec: maintain                          -333.858      47.4629    -7.03   <1e-11
load: yes                                 78.6849     16.0525     4.90   <1e-6 
spkr: old & prec: maintain               -21.8802     16.0525    -1.36   0.1729
spkr: old & load: yes                     18.3214     16.0526     1.14   0.2537
prec: maintain & load: yes                 4.47097    16.0526     0.28   0.7806
spkr: old & prec: maintain & load: yes    23.5219     16.0525     1.47   0.1428
───────────────────────────────────────────────────────────────────────────────

julia> kbsamp = parametricbootstrap(MersenneTwister(1234321), 5000, m2, use_threads=true);
Progress: 100%|█████████████████████████████████████████████████████████████████████████████████████████████| Time: 0:00:28

julia> by(DataFrame(kbsamp.allpars), [:type,:group,:names], interval = :value => shortestcovint)
13×4 DataFrame
│ Row │ type      │ group   │ names                                  │ interval             │
│     │ Symbol    │ Symbol⍰ │ Union…⍰                                │ Tuple…               │
├─────┼───────────┼─────────┼────────────────────────────────────────┼──────────────────────┤
│ 1   │ objective │ missing │ missing                                │ (28529.5, 28757.6)   │
│ 2   │ σ         │ missing │ missing                                │ (655.812, 701.061)   │
│ 3   │ β         │ missing │ (Intercept)                            │ (2035.06, 2336.22)   │
│ 4   │ β         │ missing │ spkr: old                              │ (37.4885, 97.9077)   │
│ 5   │ β         │ missing │ prec: maintain                         │ (-425.926, -241.911) │
│ 6   │ β         │ missing │ load: yes                              │ (48.0674, 111.585)   │
│ 7   │ β         │ missing │ spkr: old & prec: maintain             │ (-54.3402, 9.3307)   │
│ 8   │ β         │ missing │ spkr: old & load: yes                  │ (-11.5633, 49.9202)  │
│ 9   │ β         │ missing │ prec: maintain & load: yes             │ (-26.2791, 37.3469)  │
│ 10  │ β         │ missing │ spkr: old & prec: maintain & load: yes │ (-6.75639, 54.9565)  │
│ 11  │ σ         │ item    │ (Intercept)                            │ (264.141, 456.997)   │
│ 12  │ σ         │ item    │ prec: maintain                         │ (175.646, 316.404)   │
│ 13  │ σ         │ subj    │ (Intercept)                            │ (230.319, 356.489)   │
```

dmbates commented:
Here is where we are now.
```julia
julia> using DataFrames, MixedModels, Random

julia> sleepstudy = MixedModels.dataset(:sleepstudy);

julia> m3 = fit(MixedModel, @formula(reaction ~ 1+days+(1+days|subj)), sleepstudy);

julia> slpsamp = parametricbootstrap(MersenneTwister(42), 10_000, m3, use_threads=true);
Progress: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| Time: 0:00:03

julia> by(DataFrame(slpsamp.allpars), [:type,:group,:names], interval = :value => shortestcovint)
7×4 DataFrame
│ Row │ type      │ group   │ names                          │ interval           │
│     │ Symbol    │ Symbol⍰ │ Union…⍰                        │ Tuple…             │
├─────┼───────────┼─────────┼────────────────────────────────┼────────────────────┤
│ 1   │ objective │ missing │ missing                        │ (1709.43, 1784.75) │
│ 2   │ σ         │ missing │ missing                        │ (22.5926, 28.4618) │
│ 3   │ β         │ missing │ (Intercept)                    │ (238.698, 264.515) │
│ 4   │ β         │ missing │ days                           │ (7.62016, 13.5194) │
│ 5   │ σ         │ subj    │ (Intercept)                    │ (10.9178, 33.8073) │
│ 6   │ σ         │ subj    │ days                           │ (3.06813, 7.75216) │
│ 7   │ ρ         │ subj    │ (Symbol("(Intercept)"), :days) │ (-0.414697, 1.0)   │

julia> m3
Linear mixed model fit by maximum likelihood
 reaction ~ 1 + days + (1 + days | subj)
   logLik   -2 logLik     AIC        BIC    
 -875.96967 1751.93934 1763.93934 1783.09709

Variance components:
            Column    Variance  Std.Dev.   Corr.
subj     (Intercept)  565.51068 23.780469
         days          32.68212  5.716828  0.08
Residual              654.94145 25.591824
 Number of obs: 180; levels of grouping factors: 18

  Fixed-effects parameters:
──────────────────────────────────────────────────
             Estimate  Std.Error  z value  P(>|z|)
──────────────────────────────────────────────────
(Intercept)  251.405     6.63226    37.91   <1e-99
days          10.4673    1.50224     6.97   <1e-11
──────────────────────────────────────────────────
```
I have added the correlations to the `allpars` property.  I found in this example that I needed to be more careful in `shortestcovint` to handle non-finite values.  There are a few cases in the bootstrap where the standard deviation of one of the random effects converges to zero and the correlation is undefined.

palday commented:
> There are a few cases in the bootstrap where the standard deviation of one of the random effects converges to zero 

That's somewhat surprising for this dataset, but it's a good issue to be aware of.

For the residual variance: should `name` be `"residual"`? Having that name be `missing` somehow seems weird from a user perspective.

palday commented:
Should we also have extractors for the conditional modes? I don't think it makes sense to include them in `allpars` by default, but there are some nice use cases for having uncertainty estimates on the modes themselves.

kliegl commented:
I definitely would vote for such extractors.

dmbates commented:
Regarding getting the conditional modes.  At present the `parametricbootstrap` function returns a `RowTable` that includes the objective, sigma, beta and theta in a condensed form (the vectors are fixed-size, "static" vectors).  I plan to add the standard errors of the fixed-effects parameters to this so that the Wald tests on coefficients can be performed using bootstrap.

I think adding the conditional modes would potentially blow up the size of the returned object.  I would suggest that we document how to do a custom simulation if the user wants these.

palday commented:
Sounds good -- it might be something to add to the docs along with `condVar`.

dmbates commented:
Would it be better to have the "group" of the residual standard deviation be "residual" rather than the name?  Right now "name" is short for "column name" meaning, as in the StatsModels package.

palday commented:
That works for me.

dmbates commented:
A couple of other thoughts on format.  I have been using a `RowTable` output from `parametricbootstrap` and `ColumnTable` from the properties.  Also I have been using `Symbol`s instead of `String`s in the columns.  It may be more effective to output `ColumnTable`s in both cases and to make the descriptive columns `PooledArray{String}` or `PooledArray{Union{Missing,String}`.  The advantage of using `Symbol`s is that they are interned so it doesn't really matter if you pool them or not - they still have a compact representation because each `Symbol` is just a pointer to an entry in the symbol table.  However, they confuse operations in `DataFramesMeta`.  You can't, e.g., use `:type .== :β` because the macro wants to swallow any symbols in the expression.

palday commented on Project.toml:
```diff
@@ -27,7 +27,7 @@ Tables = "bd369af6-aec1-5ad0-b16a-f7cc5008161c"
 
 [compat]
-BlockArrays = "0.5, 0.6, 0.7, 0.8, 0.9, 0.10, 0.11"
+BlockArrays = "0.11"
```
I have no problem here, but I had to laugh a little bit because just a few months ago we were blocking on being incompatible with BlockArrays 0.11.

palday commented on src/linearmixedmodel.jl:
```diff
@@ -499,6 +499,33 @@ Test whether the model `m` is singular if the parameter vector is `θ`.
 """
 issingular(m::LinearMixedModel, θ=m.θ) = any(isapprox.(lowerbd(m), θ))
 
+function StatsBase.leverage(m::LinearMixedModel{T}) where {T}
```
This is definitely something that's nice to have, but what made you think of this here (=as part of this PR)?

palday commented on src/linearmixedmodel.jl:
```diff
@@ -779,8 +806,9 @@ function σρs(m::LinearMixedModel)
 function Base.size(m::LinearMixedModel)
-    n, p = size(first(m.feterms))
-    n, p, sum(size.(m.reterms, 2)), length(m.reterms)
+    n, p = size(fetrm(m))
```
`p` is the rank, not just the number of terms, right?

palday replied:
Using the size and not the rank here is part of the restructuring around `fixef!`, rightj?

palday commented:
I like Symbols here but if they create problems for DataFramesMeta, then we should probably not use them. Is there any sign from DataFramesMeta that they'll support symbols for such ops?

I don't know whether column or row tables make more sense here, but being consistent across the different methods is probably the most important thing.

palday commented:
@dmbates resolving the merge conflict will require a rebase .... let me know when you're done and I'll handle it before squash and merge.

dmbates commented:
I think this is ready to go. @palday could you take a look and, if it seems okay to you, pull the trigger?

palday commented:
Will do -- probably not tonight, but tomorrow sometime.

palday commented on src/bootstrap.jl:
```diff
@@ -0,0 +1,262 @@
+- `bstr`: the parameter estimates from the bootstrap replicates as a vector of named tuples.
+- `λ`: `Vector{LowerTriangular{T,Matrix{T}}}` containing copies of the λ field from `ReMat` model terms
+- `inds`: `Vector{Vector{Int}}` containing copies of the `inds` field from `ReMat` model terms
+- `lowerbd`: `Vector{T}` containing the vector of lower bounds
```
```suggestion
- `lowerbd`: `Vector{T}` containing the vector of lower bounds (corresponds to the identically named field of [`OptSummary`](@ref))
```

palday commented on src/bootstrap.jl:
```diff
@@ -0,0 +1,262 @@
+    βsc, θsc, p, k, m = similar(β), similar(θ), length(β), length(θ), deepcopy(morig)
+
+    β_names = (Symbol.(fixefnames(morig))..., )
+    rank = length(β_names)
```
This is no longer used with the new `fixef` methods.

palday commented on src/bootstrap.jl:
```diff
@@ -0,0 +1,262 @@
+    len, i = findmin([vv[i+ilen-1] - vv[i] for i = start:(stop+1-ilen)])
+    (vv[i], vv[i + ilen - 1])
+end
+
```
```suggestion

"""
    tidyβ(bsamp::MixedModelBootstrap)
Return a tidy (row)table with the parameter estimates spread into columns
of `iter`, `coefname` and `β`
"""
```

palday commented on src/bootstrap.jl:
```diff
@@ -0,0 +1,262 @@
+    end
+    result
+end
+
```
```suggestion

"""
    tidyσs(bsamp::MixedModelBootstrap)
Return a tidy (row)table with the estimates of the variance components (on the standard deviation scale) spread into columns
of `iter`, `group`, `column` and `σ`.
"""
```

palday commented on src/linearmixedmodel.jl:
```diff
@@ -466,7 +466,7 @@ function Base.getproperty(m::LinearMixedModel, s::Symbol)
         ranef(m)
     elseif s == :feterms
-        filter(Base.Fix2(isa, FeMat), getfield(m, :allterms))
+        convert(Vector{FeMat{T}}, filter(Base.Fix2(isa, FeMat), getfield(m, :allterms)))
```
This also returns the response because the response is stored as an `FeMat`. Is that the desired behavior (which is no change from previous behavior of course)?

palday replied:
Ah, yes there's also the non-exported `fetrm`.

palday commented on docs/src/bootstrap.md:
````diff
@@ -25,49 +25,45 @@ For example, a simple linear mixed-effects model for the `Dyestuff` data in the
+σres = @where(df, :type .== "σ", :group .== "residual").value
+plot(x = σres, Geom.density, Guide.xlabel("Parametric bootstrap estimates of σ"))
 ```
 For the estimates of the intercept parameter, the `getproperty` extractor must be used
````
There is no explicit call to `getproperty` here, which might be confusing to Julia neophytes (which the rest of the style seems to assume). I don't have a better suggestion at the moment, but this is a spot where we might be able to add more polish in the future.

palday commented on docs/src/bootstrap.md:
````diff
@@ -76,10 +72,22 @@ Although this mode appears to be diffuse, this is an artifact of the way that de
+```@docs
+shortestcovint
+```
+
````
```suggestion

We generate these for all random and fixed effects:
```

palday commented on docs/src/bootstrap.md:
````diff
@@ -122,22 +130,14 @@ sum(ρs .≈ -1)
-```
-
 The `issingular` method for a `LinearMixedModel` object that tests if a parameter vector `θ` corresponds to a boundary or singular fit.
 The default value of `θ` is that from the model but another value can be given as the second argument.
````
I think I would omit this line since we don't have an example here and the practical use of this isn't obvious to a typical user.
```suggestion

```

dmbates commented:
@palday - Are you okay with this PR being squashed and merged?  I would like to rebase some other branches on an updated master.

palday commented:
yes
